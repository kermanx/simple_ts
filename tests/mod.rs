use std::{fs, sync::LazyLock};

use insta::{Settings, assert_snapshot, glob};
use oxc::{
  allocator::Allocator,
  ast::{NONE, ast::Statement},
  codegen::Codegen,
  span::{SPAN, SourceType},
};
use regex::Regex;
use simple_ts::{Config, analyze};

static TYPE_QUERY_RE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"\^\? (\w+)").unwrap());

pub fn serialize_queried_types(code: String) -> String {
  let allocator = Allocator::default();
  let code = allocator.alloc_str(&code);
  let mut analyzer = analyze(&allocator, code, Config::default());
  let codegen = Codegen::new();

  let mut snapshot_stmts = analyzer.ast_builder.vec();
  for query in TYPE_QUERY_RE.find_iter(code) {
    let name = &query.as_str()[3..];
    let mark_offset = query.start();
    let mut line_col = analyzer.line_index.line_col(mark_offset.try_into().unwrap());
    line_col.line -= 1;
    let start_offset = analyzer.line_index.offset(line_col).unwrap();
    let ty = analyzer.get_type_by_pos(start_offset.into()).unwrap_or_else(|| {
      panic!("Type query `{}` at {}:{} not found", name, line_col.line + 1, line_col.col + 1)
    });
    snapshot_stmts.push(Statement::from(analyzer.ast_builder.declaration_ts_type_alias(
      SPAN,
      analyzer.ast_builder.binding_identifier(SPAN, name),
      NONE,
      analyzer.serialize_type(ty),
      false,
    )));
  }

  codegen
    .build(&analyzer.ast_builder.program(
      SPAN,
      SourceType::tsx(),
      "",
      analyzer.ast_builder.vec(),
      None,
      analyzer.ast_builder.vec(),
      snapshot_stmts,
    ))
    .code
}

#[test]
fn test() {
  glob!("fixtures/**/*.ts", |path| {
    println!("Testing {}", path.display());
    let input = fs::read_to_string(path).unwrap();
    let mut settings = Settings::clone_current();
    settings.set_omit_expression(true);
    settings.set_prepend_module_to_snapshot(false);
    settings.bind(|| {
      assert_snapshot!(serialize_queried_types(input));
    })
  });
}
