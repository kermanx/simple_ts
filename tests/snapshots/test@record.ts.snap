---
source: tests/mod.rs
input_file: tests/fixtures/record.ts
---
type Obj = {
	a: string;
	b: 111;
};
type Obj_a = string;
